### PR TITLE
Enumerate quorums method

### PR DESCRIPTION
**Pull request by liamsi**
Add enumerate quorums method and rename some methods:
- `contract_to_maximal_quorum` -> `greatest_quorum`
- prev. `is_quorum` -> `contracts_to_greatest_quorum` (used in current lachowski impl)
 
Add new methods:
- `enumerate_quorums`
- `is_quorum`
- `contains_slice`
- `traverse_quorums`

Add test for `enumerate_quorums`, move test for `is_quorum` and delete old `is_quorum` method.

### CODE REVIEW

File: stellarobservatory/quorum_lachowski_test.py [hunk 1]
```diff
@@ -71,3 +76,34 @@ def test_has_quorum_intersection_true_sccs():
...
+        5: [{4, 5, 6, 7}],
+        6: [{4, 5, 6, 7}],
+        7: [{7}],
+    }
+    quorums = list(enumerate_quorums(slices_by_node))
+    assert set(quorums) == set(
+        [frozenset({7}),
+         frozenset({4, 5, 6, 7}),
+         frozenset({1, 2, 3, 7}),
+         frozenset({1, 2, 3, 4, 5, 6, 7})]
```
Comment (liamsi) [inline stellarobservatory/quorum_lachowski_test.py:108]: This is example 2.8 (from the paper draft). I'd be curious to know how this performs with actual large FBASs.

File: stellarobservatory/quorum_lachowski.py [hunk 2]
```diff
@@ -122,19 +128,64 @@ def contract_to_maximal_quorum(nodes, slices_by_node):
...
                 filtered.add(node)
         if filtered in (nodes, {}):
             return filtered
         nodes = filtered
 
-def contains_quorum_slice(nodes_subset, slices, node):
-    """Check if for the given nodes and quorum slices there is a quorum slice
-    contained in the set of given nodes."""
+
+def is_quorum(slices, nodes_subset):
```
Comment (andrenarchy) [inline stellarobservatory/quorum_lachowski.py]: do we want to call it `slices_by_node` again? (same in `contains_slice()`)

File: stellarobservatory/quorum_lachowski.py [hunk 3]
```diff
@@ -122,19 +128,64 @@ def contract_to_maximal_quorum(nodes, slices_by_node):
...
+    else:
+        perimeter = committed.union(remaining)
+        greatest_q = greatest_quorum(perimeter, fbas_info['slices_by_node'])
+        if greatest_q == set() or not committed.issubset(greatest_q):
+            return
+        # v ← pick from R:
+        split = next_split_node(remaining, fbas_info['deps_by_node'])
+        remaining_without_split = remaining.difference({split})
+        yield from traverse_quorums(fbas_info, committed, remaining_without_split)
+        yield from traverse_quorums(fbas_info, committed.union({split}), remaining_without_split)
```
Comment (andrenarchy) [inline stellarobservatory/quorum_lachowski.py:191]: This is beautiful! Very readable and almost line by line as @torstenstueber's algorithm. :clap:

File: stellarobservatory/quorum_lachowski_test.py (hunk 1 above)
Comment (andrenarchy) [reply stellarobservatory/quorum_lachowski_test.py:108]: We'll find out! :sunglasses:

**Review by andrenarchy — APPROVED** [review]

**Comment by andrenarchy** [comment]
Timings for the following quorum slice definition (with Stellar Core style postprocessing for each node):
```
{
    'threshold': 4,
    'validators': [],
    'innerQuorumSets': [
        {'threshold': 2, 'validators': ['coinqvest-de', 'coinqvest-fi', 'coinqvest-hk'], 'innerQuorumSets': []},
        {'threshold': 2, 'validators': ['satoshipay-de-fra', 'satoshipay-sg-sin', 'satoshipay-us-iowa'], 'innerQuorumSets': []},
        {'threshold': 2, 'validators': ['sdf1', 'sdf2', 'sdf3'], 'innerQuorumSets': []},
        {'threshold': 3, 'validators': ['lobstr1', 'lobstr2', 'lobstr3', 'lobstr4', 'lobstr5'], 'innerQuorumSets': []},
        {'threshold': 2, 'validators': ['keybase0', 'keybase1', 'keybase2'], 'innerQuorumSets': []},
    ]
}
```
Lachowski Algorithm: 34s
Powerset iteration: 48s

**Pull request merged**